include/hydrosheds/hydrosheds.hh

```diff
@@ -4,59 +4,193 @@
 
 #include<array>
 #include<string>
+#include <cmath>
 
 namespace hydrosheds {
 
-  using Coordinate = std::array<double, 2>;
-  using HydroshedsID = int;
+    using Coordinate = std::array<double, 2>;
+    using HydroshedsID = int;
 
-  // Forward declarations
-  class RiverSegment;
-  class FullDatasetRiverSegmentIterator;
+    // Forward declarations
+    class RiverSegment;
+    class FullDatasetRiverSegmentIterator;
 
-  class HydroshedsDataSet
-  {
+    class DownstreamIterator;
+    class DownstreamIteratorHelper;
+
+
+    class HydroshedsDataSet
+    {
     public:
-    HydroshedsDataSet(const std::string&);
+        HydroshedsDataSet(const std::string&);
+
+        // Iterator access for Feature Iteration
 
-    // Iterator access
-    FullDatasetRiverSegmentIterator begin() const;
-    FullDatasetRiverSegmentIterator end() const;
+        FullDatasetRiverSegmentIterator begin() const;
+        FullDatasetRiverSegmentIterator end() const;
 
-    // Not intended to stay - just get some segment
-    RiverSegment getSegment() const;
+        // Iterator access for Segment Iteration
+        DownstreamIteratorHelper followbegin();
+        DownstreamIteratorHelper followend();
+
+        // Not intended to stay - just get some segment
+        RiverSegment getSegment();
 
     private:
-    OGRLayer* layer;
-  };
+        friend class DownstreamIteratorHelper;
+        friend class DownstreamIterator;
+        friend class RiverSegment;
+        friend class FullDatasetRiverSegmentIterator;
+        OGRLayer* layer;
+        long count;
+    };
 
 
-  class RiverSegment
-  {
+    class RiverSegment
+    {
     public:
-    double getLength() const;
-    double getGeologicalLength() const;
-    double getDischarge() const;
+        double getLength() const;
+        double getGeologicalLength() const;
+        double getDischarge() const;
+
+        Coordinate getStartingPoint() const;
+        Coordinate getEndPoint() const;
+
+        OGRFeature* getFeature(){return feature;};
+        OGRLayer* getLayer(){return layer;};
```

Comment:
What do we need these for when we also have a number of `friend` declarations in place?

Reply:
I ran into trouble while implementing these as a friend declarations. So I implemented these helper methods.

Reply:
Actually there still occurs some troubles using these friend declarations. Don't know exactly what causes this, but I am trying to fix that.